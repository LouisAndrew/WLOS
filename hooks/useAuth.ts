import { useContext } from 'react'
import { AuthContext } from '@lib/context/AuthContext'

export const useAuth = () => useContext(AuthContext)
