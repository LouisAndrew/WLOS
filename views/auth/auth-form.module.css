.wrapper {
  @apply flex flex-col justify-center h-full;
}

/* .pass-icon {
  fill: theme('colors.primary-yellow');
  @apply absolute bottom-4 right-4 h-6 w-6 cursor-pointer;
} */

.breadcrumb {
  @apply text-shade-gray font-display text-sm tracking-extrawide mt-3;
}

.button {
  @apply w-full justify-center text-base py-3 md:text-sm;
}

.password-toggle {
  @apply absolute bottom-4 right-4 cursor-pointer md:bottom-3 text-primary-yellow duration-200
    opacity-0;
}

.password-toggle.show {
  @apply opacity-100;
}
