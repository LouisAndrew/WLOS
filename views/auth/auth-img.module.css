.wrapper {
  @apply relative bg-bg-gray flex flex-center;
}

.background-wrapper {
  @apply absolute z-10 w-full h-full top-0 left-0 px-6
    flex flex-col items-center justify-center;
}

.yellow-stripe {
  transform: skew(-15deg) rotate(-5deg);
  @apply h-24 w-full bg-primary-yellow;
}

.white-stripe {
  transform: skew(-12deg) rotate(-5deg);
  @apply h-28 w-4/5 bg-white mt-8;
}

.img-wrapper {
  @apply max-w-xs relative z-20 w-3/4;
}
