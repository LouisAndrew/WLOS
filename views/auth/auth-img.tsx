import React, { FC } from 'react'
import Image from 'next/image'

const imgSrc = '/auth-athlete-img.png'
import styles from './auth-img.module.css'

export type Props = {}

const AuthImage: FC<Props> = () => {
  return (
    <div data-testid="wrapper" className={styles.wrapper}>
      <div className={styles['background-wrapper']}>
        <div className={styles['yellow-stripe']} />
        <div className={styles['white-stripe']} />
      </div>
      <div className={styles['img-wrapper']}>
        <Image src={imgSrc} width={805} height={1468} layout="responsive" />
      </div>
    </div>
  )
}

export default AuthImage
