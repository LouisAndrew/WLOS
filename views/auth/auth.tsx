import React, { FC, useState } from 'react'

import classname from 'classnames/bind'

import styles from './auth.module.css'
import AuthImage from './auth-img'
import AuthForm from './auth-form'

const cx = classname.bind(styles)

export type Props = {}

const Auth: FC<Props> = () => {
  const [displayAuthImage, setDisplayAuthImage] = useState(true)

  const quote = 'Millions of people use this app to make their life better'

  const authFormWrapperClass = cx({
    'auth-form__wrapper': true,
    'display-form': !displayAuthImage,
  })

  return (
    <div data-testid="auth-wrapper" className={`h-full-screen ${styles.wrapper}`}>
      <div className={styles['auth-image__wrapper']}>
        <AuthImage />
        <h5>Get Started</h5>
        <h3 className={styles['auth-image__heading']}>{quote}</h3>
        <button
          className={`btn btn--primary ${styles['auth-image__next-btn']}`}
          onClick={() => setDisplayAuthImage(false)}
        >
          Next
        </button>
      </div>
      <div className={authFormWrapperClass}>
        <AuthForm />
      </div>
    </div>
  )
}

export default Auth
