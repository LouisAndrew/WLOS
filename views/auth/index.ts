import Auth from './auth'
import AuthImage from './auth-img'

export { Auth, AuthImage }
