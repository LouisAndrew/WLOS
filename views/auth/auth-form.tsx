import { useRouter } from 'next/router'
import React, { FC, useEffect, useState } from 'react'
import { useAuth } from '@h/useAuth'
import classname from 'classnames/bind'

import styles from './auth-form.module.css'

const cx = classname.bind(styles)

export type Props = {}

type AuthState = 'SIGN_IN' | 'SIGN_UP'

const texts: Record<AuthState, { btnText: string; headingText: string }> = {
  SIGN_IN: {
    headingText: 'If you are new',
    btnText: 'Create New Account',
  },
  SIGN_UP: {
    headingText: 'Already have an account?',
    btnText: 'Sign in',
  },
}

const AuthForm: FC<Props> = () => {
  const router = useRouter()
  const { signIn, signInWithGoogle, signUp } = useAuth()
  const [authState, setAuthState] = useState<AuthState>('SIGN_IN')
  const [formState, setFormState] = useState<{ username: string; password: string }>({
    username: '',
    password: '',
  })
  const [showPassword, setShowPassword] = useState(false)

  const headingText = authState === 'SIGN_IN' ? 'Login Now.' : 'Create an account'

  useEffect(() => {
    const path = router.asPath.replace('/auth?', '')

    if (path === 'sign-up') {
      setAuthState('SIGN_UP')
      return
    }
    setAuthState('SIGN_IN')
  }, [])

  useEffect(() => {
    router.replace(`auth?${authState.toLowerCase().split('_').join('-')}`, undefined, {
      shallow: true,
    })
  }, [authState])

  const handleClickSignInGoogle = async () => {
    await signInWithGoogle()
  }
  // todo: on success state
  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    const { username, password } = formState
    if (username && password) {
      if (authState === 'SIGN_IN') {
        await signIn(username, password)
      } else {
        await signUp(username, password)
      }
    }
  }

  const passwordToggleClass = cx({
    'password-toggle': true,
    show: formState.password,
  })

  return (
    <div data-testid="auth-form-wrapper" className={styles.wrapper}>
      <h1>
        Hey,
        <span className="text-primary-yellow block">{headingText}</span>
      </h1>
      <div className={styles.breadcrumb}>
        {texts[authState].headingText} /
        <button
          className="btn btn--secondary btn--text inline-block"
          onClick={() => setAuthState(authState === 'SIGN_IN' ? 'SIGN_UP' : 'SIGN_IN')}
        >
          {texts[authState].btnText}
        </button>
      </div>
      <form onSubmit={handleSubmit} role="form" className="py-3 duration-200">
        <label htmlFor="username">
          Username
          <input
            type="text"
            id="username"
            className="input input--primary"
            placeholder="Username"
            value={formState.username}
            onChange={(e) => setFormState((prev) => ({ ...prev, username: e.target.value }))}
          />
        </label>
        <label htmlFor="password" className="block mt-6 relative">
          Password
          <input
            type={showPassword ? 'text' : 'password'}
            id="password"
            className="input input--secondary pr-20"
            placeholder="Password"
            value={formState.password}
            onChange={(e) => setFormState((prev) => ({ ...prev, password: e.target.value }))}
          />
          <span onClick={() => setShowPassword((prev) => !prev)} className={passwordToggleClass}>
            {showPassword ? 'HIDE' : 'SHOW'}
          </span>
        </label>
        {authState === 'SIGN_IN' ? (
          <>
            <button
              data-testid="sign-in"
              type="submit"
              className={`btn btn--primary mb-3 mt-12 ${styles.button}`}
            >
              Sign In
            </button>
            <button
              data-testid="sign-in-google"
              className={`btn btn--secondary ${styles.button}`}
              onClick={handleClickSignInGoogle}
            >
              Sign In with Google
            </button>
          </>
        ) : (
          <button
            data-testid="sign-up"
            type="submit"
            className={`btn btn--primary mb-3 mt-12 ${styles.button}`}
          >
            Sign Up
          </button>
        )}
      </form>
    </div>
  )
}

export default AuthForm
