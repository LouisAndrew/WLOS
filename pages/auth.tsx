import { Auth as AuthView } from '../views/auth'

export default function Auth() {
  return <AuthView />
}
