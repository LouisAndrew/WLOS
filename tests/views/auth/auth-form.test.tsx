import { render, cleanup, fireEvent } from '@testing-library/react'
import * as auth from '@h/useAuth'

import { AuthForm } from '@v/auth'
import { mockNextUseRouter } from '@tests/utils/mockNextRouter'
import { mockUseAuth } from '@tests/utils/mockUseAuth'

const signIn = jest.fn()
const signInWithGoogle = jest.fn()
const signUp = jest.fn()

afterEach(cleanup)
beforeEach(() => {
  mockNextUseRouter({
    asPath: '/auth?sign-in',
  })
})

const testIds = {
  signIn: 'sign-in',
  signInGoogle: 'sign-in-google',
  signUp: 'sign-up',
}

const username = 'john@doe.com'
const password = 123456

const Component = <AuthForm />

describe('Auth form component', () => {
  it('renders the component', () => {
    const wrapper = render(Component)
    expect(wrapper.queryByTestId('auth-form-wrapper')).toBeInTheDocument()
  })

  it('matches the snapshot', () => {
    expect(render(Component)).toMatchSnapshot()
  })

  describe('renders appropriate buttons on sign-in', () => {
    beforeEach(() => {
      mockNextUseRouter({
        query: { 'sign-in': '' },
      })
      mockUseAuth({
        signIn,
        signInWithGoogle,
        signUp,
      })
    })

    afterEach(cleanup)
    // const { queryByTestId: q, getByLabelText: g, getByTestId: gTid, debug } = render(Component)
    it('renders the sign in button', () => {
      expect(render(Component).queryByTestId(testIds.signIn)).toBeInTheDocument()
    })

    it('renders the signin with google button', () => {
      expect(render(Component).queryByTestId(testIds.signInGoogle)).toBeInTheDocument()
    })

    it('does not render the sign up button', () => {
      expect(render(Component).queryByTestId(testIds.signUp)).not.toBeInTheDocument()
    })

    it('calls the `signIn` button when sign in button is clicked and inputs are filled', () => {
      const { getByLabelText: g, getByRole: gRole, debug, getByPlaceholderText: gptxt } = render(
        Component
      )

      const usernameInput = g('Username')
      const passwordInput = gptxt('Password')

      fireEvent.change(usernameInput, { target: { value: username } })
      fireEvent.change(passwordInput, { target: { value: password } })
      // fireEvent.click(gTid(testIds.signIn))
      fireEvent.submit(gRole('form'))

      // event propagation not supported
      setTimeout(() => {
        expect(signIn).toBeCalled()
        expect(signIn).toBeCalledWith(username, password)
      }, 50)
    })

    it('does not call the `signIn` function when the inputs are not filled completely', () => {
      const { getByTestId: gTid } = render(Component)
      fireEvent.click(gTid(testIds.signIn))
      expect(signIn).not.toBeCalled()
    })

    it('calls the `signInWithGoogle` function when sign-in-google button is clicked', () => {
      const { getByTestId: gTid } = render(Component)
      fireEvent.click(gTid(testIds.signInGoogle))
      setTimeout(() => {
        expect(signInWithGoogle).toBeCalled()
        expect(signInWithGoogle).toBeCalledWith()
      }, 50)
    })
  })

  describe('renders appropriate buttons on sign-up', () => {
    beforeEach(() => {
      mockNextUseRouter({ asPath: '/auth?sign-up' })
      mockUseAuth({
        signIn,
        signInWithGoogle,
        signUp,
      })
    })

    afterEach(cleanup)
    // const { queryByTestId: q, getByTestId: gTid, getByLabelText: g } = render(Component)
    it('does not render the sign in button', () => {
      const { queryByTestId: q } = render(Component)

      expect(q(testIds.signIn)).not.toBeInTheDocument()
    })

    it('does not render the signin with google button', () => {
      const { queryByTestId: q } = render(Component)

      expect(q(testIds.signInGoogle)).not.toBeInTheDocument()
    })

    it('renders the sign up button', () => {
      const { queryByTestId: q } = render(Component)

      expect(q(testIds.signUp)).toBeInTheDocument()
    })

    it('calls the `signIn` button when sign in button is clicked and inputs are filled', () => {
      const { getByRole: gRole, getByLabelText: g, getByPlaceholderText: gptxt } = render(Component)
      const usernameInput = g('Username')
      const passwordInput = gptxt('Password')

      fireEvent.change(usernameInput, { target: { value: username } })
      fireEvent.change(passwordInput, { target: { value: password } })
      fireEvent.click(gRole('form'))

      setTimeout(() => {
        expect(signUp).toBeCalled()
        expect(signUp).toBeCalledWith(username, password)
      }, 50)
    })

    it('does not call the `signUp` function when the inputs are not filled completely', () => {
      const { getByTestId: gTid } = render(Component)

      fireEvent.click(gTid(testIds.signUp))
      expect(signUp).not.toBeCalled()
    })
  })
})
