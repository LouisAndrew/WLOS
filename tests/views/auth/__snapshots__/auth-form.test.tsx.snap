// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth form component matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrapper"
        data-testid="auth-form-wrapper"
      >
        <h1>
          Hey,
          <span
            class="text-primary-yellow block"
          >
            Login Now.
          </span>
        </h1>
        <div
          class="breadcrumb"
        >
          If you are new
           /
          <button
            class="btn btn--secondary btn--text inline-block"
          >
            Create New Account
          </button>
        </div>
        <form
          class="py-3 duration-200"
          role="form"
        >
          <label
            for="username"
          >
            Username
            <input
              class="input input--primary"
              id="username"
              placeholder="Username"
              type="text"
              value=""
            />
          </label>
          <label
            class="block mt-6 relative"
            for="password"
          >
            Password
            <input
              class="input input--secondary pr-20"
              id="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <span
              class="password-toggle"
            >
              SHOW
            </span>
          </label>
          <button
            class="btn btn--primary mb-3 mt-12 button"
            data-testid="sign-in"
            type="submit"
          >
            Sign In
          </button>
          <button
            class="btn btn--secondary button"
            data-testid="sign-in-google"
          >
            Sign In with Google
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrapper"
      data-testid="auth-form-wrapper"
    >
      <h1>
        Hey,
        <span
          class="text-primary-yellow block"
        >
          Login Now.
        </span>
      </h1>
      <div
        class="breadcrumb"
      >
        If you are new
         /
        <button
          class="btn btn--secondary btn--text inline-block"
        >
          Create New Account
        </button>
      </div>
      <form
        class="py-3 duration-200"
        role="form"
      >
        <label
          for="username"
        >
          Username
          <input
            class="input input--primary"
            id="username"
            placeholder="Username"
            type="text"
            value=""
          />
        </label>
        <label
          class="block mt-6 relative"
          for="password"
        >
          Password
          <input
            class="input input--secondary pr-20"
            id="password"
            placeholder="Password"
            type="password"
            value=""
          />
          <span
            class="password-toggle"
          >
            SHOW
          </span>
        </label>
        <button
          class="btn btn--primary mb-3 mt-12 button"
          data-testid="sign-in"
          type="submit"
        >
          Sign In
        </button>
        <button
          class="btn btn--secondary button"
          data-testid="sign-in-google"
        >
          Sign In with Google
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
